src/spaces/jolt_physics_direct_space_state_3d.cpp

```diff
@@ -983,28 +1021,73 @@ bool JoltPhysicsDirectSpaceState3D::_body_motion_collide(
 		const JoltShapedObjectImpl3D* collider = collider_jolt_body.as_shaped();
 		ERR_FAIL_NULL_D(collider);
 
-		const Vector3 position = base_offset + to_godot(hit.mContactPointOn2);
-
 		const int32_t local_shape = p_body.find_shape_index(hit.mSubShapeID1);
 		ERR_FAIL_COND_D(local_shape == -1);
 
 		const int32_t collider_shape = collider->find_shape_index(hit.mSubShapeID2);
 		ERR_FAIL_COND_D(collider_shape == -1);
 
-		PhysicsServer3DExtensionMotionCollision& collision = p_result->collisions[count++];
-
-		collision.position = position;
-		collision.normal = to_godot(-hit.mPenetrationAxis.Normalized());
-		collision.collider_velocity = collider->get_velocity_at_position(position);
-		collision.collider_angular_velocity = collider->get_angular_velocity();
-		collision.depth = penetration_depth;
-		collision.local_shape = local_shape;
-		collision.collider_id = collider->get_instance_id();
-		collision.collider = collider->get_rid();
-		collision.collider_shape = collider_shape;
+		for (JPH::Vec3 contact_point : contact_points2) {
+			const Vector3 position = base_offset + to_godot(contact_point);
+
+			PhysicsServer3DExtensionMotionCollision& collision = p_result->collisions[count++];
+
+			collision.position = position;
+			collision.normal = to_godot(-hit.mPenetrationAxis.Normalized());
+			collision.collider_velocity = collider->get_velocity_at_position(position);
+			collision.collider_angular_velocity = collider->get_angular_velocity();
+			collision.depth = penetration_depth;
+			collision.local_shape = local_shape;
+			collision.collider_id = collider->get_instance_id();
+			collision.collider = collider->get_rid();
+			collision.collider_shape = collider_shape;
+
+			if (count == p_max_collisions) {
+				break;
+			}
+		}
+
+		if (count == p_max_collisions) {
+			break;
+		}
 	}
 
 	p_result->collision_count = count;
 
 	return count > 0;
 }
+
+void JoltPhysicsDirectSpaceState3D::_generate_manifold(
+	const JPH::CollideShapeResult& p_hit,
+	const JPH::CollideShapeSettings& p_settings,
+	JPH::ContactPoints& p_contact_points1,
+	JPH::ContactPoints& p_contact_points2
+#ifdef JPH_DEBUG_RENDERER
+	,
+	JPH::RVec3Arg p_center_of_mass
+#endif // JPH_DEBUG_RENDERER
+) {
+	JPH::ManifoldBetweenTwoFaces(
+		p_hit.mContactPointOn1,
+		p_hit.mContactPointOn2,
+		p_hit.mPenetrationAxis,
+		p_settings.mMaxSeparationDistance,
+		p_hit.mShape1Face,
+		p_hit.mShape2Face,
+		p_contact_points1,
+		p_contact_points2
+#ifdef JPH_DEBUG_RENDERER
+		,
+		p_center_of_mass
+#endif // JPH_DEBUG_RENDERER
+	);
+
+	if (p_contact_points1.size() > 4) {
+		JPH::PruneContactPoints(
+			p_hit.mPenetrationAxis,
+			p_contact_points1,
+			p_contact_points2,
+			p_center_of_mass
+		);
+	}
```

Comment:
@jrouwe Would you say this is ill-advised? I noticed you used 32 points as the cut-off in your own usage of this, but I figured if it's not outrageously expensive it might be a good thing to do for most manifolds.

Reply:
The 32 points is just for intermediate results. In PhysicsSystem::ProcessBodyPair it keeps accumulating contact points from different hits with the same normal until the buffer is full (64 is full, you get max 32 per hit so it prunes them if there are more than 32 in the buffer currently to ensure there's enough space). When all hits have been processed it does a final prune to limit it to 4:

https://github.com/jrouwe/JoltPhysics/blob/e60de1aa544dbdb77916b5ef410f2563e2b313bd/Jolt/Physics/PhysicsSystem.cpp#L1138-L1139

It's hard to see from this diff exactly what you are doing, but if you're only pruning once then 4 should be good. If you're combining many hits and run prune multiple times then you may want to do what I do to limit the amount of calls to prune.

Reply:
Ah, gotcha. There's only one prune, so 4 it is then. Thank you!